wandb/apis/public/artifacts.py

```diff
@@ -528,21 +528,216 @@ def description(self):
+                    input: {
+                        artifactSequenceID: $artifactSequenceID
+                        name: $name
+                        description: $description
+                    }
+                    ) {
+                    artifactCollection {
+                        id
+                        name
+                        description
+                    }
+                    }
+                }
+            """)
+
+            variable_values = {
+                "artifactSequenceID": self.id,
+                "name": self._name,
+                "description": self.description,
+            }
+            self.client.execute(mutation, variable_values=variable_values)
+            self._saved_name = self._name
+
+            if self._saved_type != self._type:
+                type_mutation = gql("""
+                    mutation MoveArtifactCollection(
+                        $artifactSequenceID: ID!
+                        $destinationArtifactTypeName: String!
+                    ) {
+                        moveArtifactSequence(
+                        input: {
+                            artifactSequenceID: $artifactSequenceID
+                            destinationArtifactTypeName: $destinationArtifactTypeName
+                        }
+                        ) {
+                        artifactCollection {
+                            id
+                            name
+                            description
+                            __typename
+                        }
+                        }
+                    }
+                    """)
+
+                variable_values = {
+                    "artifactSequenceID": self.id,
+                    "destinationArtifactTypeName": self._type,
+                }
+                self.client.execute(type_mutation, variable_values=variable_values)
+                self._saved_type = self._type
+        else:
+            mutation = gql("""
+                mutation UpdateArtifactPortfolio(
+                    $artifactPortfolioID: ID!
+                    $name: String
+                    $description: String
+                ) {
+                    updateArtifactPortfolio(
+                    input: {
+                        artifactPortfolioID: $artifactPortfolioID
+                        name: $name
+                        description: $description
+                    }
+                    ) {
+                    artifactCollection {
+                        id
+                        name
+                        description
+                    }
+                    }
+                }
+            """)
+            variable_values = {
+                "artifactPortfolioID": self.id,
+                "name": self._name,
+                "description": self.description,
+            }
+            self.client.execute(mutation, variable_values=variable_values)
+            self._saved_name = self._name
```

Comment:
It seems like for all of these operations above, we don't need to keep the old type or name in order to execute the query, just like you don't keep the old description. It seems like the main advantage is that we can avoid doing sending some mutations if we know e.g. that the type hasn't changed? Though that only looks to be the case for `type` specifically.

I'm not strongly opinionated on whether we should simplify the code and assume that `save` always updates, or whether we should try to avoid round trips--normally I would be happy to accept some code complexity to improve performance, but it feels more reasonable to have `ArtifactCollection.save` take an extra 100ms than most other functions. It shouldn't be common and it's obvious it's going to require backend communication.

That's a long way of saying that it's up to you, but highlighting that it's a choice here.

Reply:
the main reason I kept `_saved_type` and `_saved_name` is because they're needed for `ArtifactCollection.artifacts`, not really sure how to get around this 😕